src/SessionWidget.vala

```diff
@@ -55,27 +67,52 @@ public class A11Y.SessionWidget : Gtk.Grid {
 
         var hover_click = new Wingpanel.Widgets.Switch (_("Dwell Click"));
 
-        var settings_button = new Gtk.ModelButton ();
-        settings_button.text = _("Universal Access Settings…");
-
         orientation = Gtk.Orientation.VERTICAL;
         add (font_size_grid);
         add (screen_reader);
         add (onscreen_keyboard);
+        add (new Wingpanel.Widgets.Separator ());
         add (slow_keys);
         add (bounce_keys);
         add (sticky_keys);
         add (hover_click);
-        add (new Wingpanel.Widgets.Separator ());
-        add (settings_button);
 
-        settings_button.clicked.connect (() => {
+        if (server_type == Wingpanel.IndicatorManager.ServerType.SESSION) {
+            var settings_button = new Gtk.ModelButton ();
+            settings_button.text = _("Universal Access Settings…");
+
+            add (new Wingpanel.Widgets.Separator ());
+            add (settings_button);
+
+            settings_button.clicked.connect (() => {
+                try {
+                    AppInfo.launch_default_for_uri ("settings://universal-access", null);
+                } catch (Error e) {
+                    warning ("Failed to open universal access settings: %s", e.message);
+                }
+            });
+
+            var applications_settings = new Settings ("org.gnome.desktop.a11y.applications");
+            applications_settings.bind ("screen-keyboard-enabled", onscreen_keyboard, "active", SettingsBindFlags.DEFAULT);
+            applications_settings.bind ("screen-reader-enabled", screen_reader, "active", SettingsBindFlags.DEFAULT);
+        } else {
+            settings = new KeyFile ();
+            settings.set_boolean ("greeter", "onscreen-keyboard", false);
+
+            screen_reader.notify["active"].connect (() => {
+                toggle_screen_reader (screen_reader.active);
+            });
+
+            onscreen_keyboard.notify["active"].connect (() => {
+                toggle_keyboard (onscreen_keyboard.active);
+            });
+
             try {
-                AppInfo.launch_default_for_uri ("settings://universal-access", null);
+                onscreen_keyboard.active = settings.get_boolean ("greeter", "onscreen-keyboard");
```

Comment:
What is the point of this? You just set that setting to "false" a few lines back.  Can't you just set the keyboard switch to `off` directly, with out a try/catch clause?